JIT ARM64: Add IF_SVE_DT_3A, IF_SVE_DU_3A, IF_SVE_DX_3A, IF_SVE_DY_3A

amanasifkhalid opened the pull request:
Part of #94549. cc @dotnet/arm64-contrib, @a74nh.

JitDisasm output:
```
whilege p0.b, w0, w1
whilege p1.b, x2, x3
whilegt p2.b, w4, w5
whilegt p3.b, x6, x7
whilehi p4.h, w8, w9
whilehi p5.h, x10, x11
whilehs p6.h, w12, w13
whilehs p7.h, x14, x15
whilele p8.s, w0, w1
whilele p9.s, x2, x3
whilelo p10.s, w4, w5
whilelo p11.s, x6, x7
whilels p12.d, w8, w9
whilels p13.d, x10, x11
whilelt p14.d, w12, w13
whilelt p15.d, x14, x15
whilerw p0.b, x0, x1
whilerw p1.h, x2, x3
whilerw p2.s, x4, x5
whilerw p3.d, x6, x7
whilewr p4.b, x8, x9
whilewr p5.h, x10, x11
whilewr p6.s, x12, x13
whilewr p7.d, x14, x15
whilege { p0.b, p1.b }, x0, x1
whilegt { p2.b, p3.b }, x2, x3
whilehi { p4.h, p5.h }, x4, x5
whilehs { p6.h, p7.h }, x6, x7
whilele { p8.s, p9.s }, x8, x9
whilelo { p10.s, p11.s }, x10, x11
whilels { p12.d, p13.d }, x12, x13
whilelt { p14.d, p15.d }, x14, x15
whilege p8.b, x0, x1, vlx2
whilege p9.b, x2, x3, vlx4
whilegt p10.h, x4, x5, vlx2
whilegt p11.h, x6, x7, vlx4
whilehi p12.s, x8, x9, vlx2
whilehi p13.s, x10, x11, vlx4
whilehs p14.d, x12, x13, vlx2
whilehs p15.d, x14, x15, vlx4
whilele p8.b, x0, x1, vlx2
whilele p9.b, x2, x3, vlx4
whilelo p10.h, x4, x5, vlx2
whilelo p11.h, x6, x7, vlx4
whilels p12.d, x8, x9, vlx2
whilels p13.d, x10, x11, vlx4
whilelt p14.b, x12, x13, vlx2
whilelt p15.b, x14, x15, vlx4
```

cstool output:

```
00002125  whilege       p0.b, w0, w1
41102325  whilege       p1.b, x2, x3
92002525  whilegt       p2.b, w4, w5
D3102725  whilegt       p3.b, x6, x7
14096925  whilehi       p4.h, w8, w9
55196B25  whilehi       p5.h, x10, x11
86096D25  whilehs       p6.h, w12, w13
C7196F25  whilehs       p7.h, x14, x15
1804A125  whilele       p8.s, w0, w1
5914A325  whilele       p9.s, x2, x3
8A0CA525  whilelo       p10.s, w4, w5
CB1CA725  whilelo       p11.s, x6, x7
1C0DE925  whilels       p12.d, w8, w9
5D1DEB25  whilels       p13.d, x10, x11
8E05ED25  whilelt       p14.d, w12, w13
CF15EF25  whilelt       p15.d, x14, x15
10302125  whilerw       p0.b, x0, x1
51306325  whilerw       p1.h, x2, x3
9230A525  whilerw       p2.s, x4, x5
D330E725  whilerw       p3.d, x6, x7
04312925  whilewr       p4.b, x8, x9
45316B25  whilewr       p5.h, x10, x11
8631AD25  whilewr       p6.s, x12, x13
C731EF25  whilewr       p7.d, x14, x15
10502125  whilege       { p0.b, p1.b }, x0, x1
53502325  whilegt       { p2.b, p3.b }, x2, x3
95586525  whilehi       { p4.h, p5.h }, x4, x5
D6586725  whilehs       { p6.h, p7.h }, x6, x7
1955A925  whilele       { p8.s, p9.s }, x8, x9
5A5DAB25  whilelo       { p10.s, p11.s }, x10, x11
9D5DED25  whilels       { p12.d, p13.d }, x12, x13
DE55EF25  whilelt       { p14.d, p15.d }, x14, x15
18402125  whilegt       pn8.b, x0, x1, vlx2
59602325  whilegt       pn9.b, x2, x3, vlx4
9A406525  whilegt       pn0xA.h, x4, x5, vlx2
DB606725  whilegt       pn0xB.h, x6, x7, vlx4
1C49A925  whilehi       pn0xC.s, x8, x9, vlx2
5D69AB25  whilehi       pn0xD.s, x10, x11, vlx4
9E49ED25  whilehi       pn0xE.d, x12, x13, vlx2
DF69EF25  whilehi       pn0xF.d, x14, x15, vlx4
18442125  whilele       pn8.b, x0, x1, vlx2
59642325  whilele       pn9.b, x2, x3, vlx4
9A4C6525  whilels       pn0xA.h, x4, x5, vlx2
DB6C6725  whilels       pn0xB.h, x6, x7, vlx4
1C4DE925  whilels       pn0xC.d, x8, x9, vlx2
5D6DEB25  whilels       pn0xD.d, x10, x11, vlx4
9E452D25  whilele       pn0xE.b, x12, x13, vlx2
DF652F25  whilele       pn0xF.b, x14, x15, vlx4
```

Notice that for the last few instructions with vector length specifiers, cstool prints the predicate number slightly differently from how we currently do it in the JIT (e.g. `pn0xA.h` vs `p10.h`). Should we match cstool's formatting exactly, or is this unimportant in this case?

amanasifkhalid committed:
Add IF_SVE_DT_3A/IF_SVE_DU_3A/IF_SVE_DX_3A/IF_SVE_DY_3A

amanasifkhalid committed:
Merge from main

amanasifkhalid committed:
Print vl

amanasifkhalid committed:
Add insOptsScalableWithVectorLength to insOptsScalable

amanasifkhalid committed:
Style

kunalspathak commented:
> Notice that for the last few instructions with vector length specifiers, cstool prints the predicate number slightly differently from how we currently do it in the JIT (e.g. `pn0xA.h` vs `p10.h`). Should we match cstool's formatting exactly, or is this unimportant in this case?

I don't think cstool output is correct. They are printing the register name in hex. @TIHan something to fix.

kunalspathak commented:
I am seeing some differences in the instruction generated. Perhaps cstool uses alias and we should probably do that too:
left = coreclr
right = cstool

<img width="1404" alt="image" src="https://github.com/dotnet/runtime/assets/12488060/eb2aecff-a48b-414c-b143-7af64e041922">

amanasifkhalid commented:
> Perhaps cstool uses alias and we should probably do that too

I can update my changes to use the same aliases as cstool. I'm guessing we'd have to split some of the `INST3` groups in `instrsarm64sve.h` up into separate groups to get the JIT to use different display names depending on the encoding, right? If we do it this way, then I might be able to remove some of the new `insOpts` I added in this change, as they are causing issues, too. The failing pipelines have to do with the `insOpts` enum having too many entries to fit in an `unsigned` type, so trying to set `_idInsOpt` to something like `INS_OPTS_ALIGN` (the "largest" `insOpts` value) results in a narrowing cast that changes `_idInsOpt` to `INS_OPTS_UXTW`, thus triggering asserts.

a74nh commented:
> > Perhaps cstool uses alias and we should probably do that too
> 
> I can update my changes to use the same aliases as cstool. I'm guessing we'd have to split some of the `INST3` groups in `instrsarm64sve.h` up into separate groups to get the JIT to use different display names depending on the encoding, right? If we do it this way, then I might be able to remove some of the new `insOpts` I added in this change, as they are causing issues, too. The failing pipelines have to do with the `insOpts` enum having too many entries to fit in an `unsigned` type, so trying to set `_idInsOpt` to something like `INS_OPTS_ALIGN` (the "largest" `insOpts` value) results in a narrowing cast that changes `_idInsOpt` to `INS_OPTS_UXTW`, thus triggering asserts.

Generally we should be using the preferred alias, as it's intended as easier to read.

What we've been doing for other instructions with aliases is just to change `ins` within the `emitIns_()` functions.

For `cpy -> mov` it's an easy switch:
https://github.com/dotnet/runtime/blob/8bd23f0792c6339478222ebc6635266e8f4098b4/src/coreclr/jit/emitarm64.cpp#L8842

For `cmpXX` instructions, it was easier to add an extra switch at the end of the main switch:
https://github.com/dotnet/runtime/blob/8bd23f0792c6339478222ebc6635266e8f4098b4/src/coreclr/jit/emitarm64.cpp#L9987

a74nh commented:
However, having said that about aliases, I don't think that's the issue here.

The difference in the output is, for example, `WHILEGE` -> `WHILEGT`. Those two can't be aliases of each other. Plus there is no mention of alases for these instructions in the manuals.

Looking at `WHILEGE` and `WHILEGT`, the only difference between the encoding of them is that bit 3 (`eq`) is set on `WHILEGE`:
https://docsmirror.github.io/A64/2023-06/whilege_pn_rr.html
https://docsmirror.github.io/A64/2023-06/whilegt_pn_rr.html

Which suggests something is going wrong in the coreclr encoding.

Looking at `PNd`, this is a 3bit field which is used to represent the registers `p8` to `p15`. ie: `000` would be `v8`, `001` would be `v9` etc.

`insEncodeReg_P_2_to_0()` is wrong. It should be something like:
```csharp
/*static*/ emitter::code_t emitter::insEncodeReg_P_2_to_0(regNumber reg)
{
    assert(isPredicateRegister(reg));
    emitter::code_t ureg = (emitter::code_t)reg - (emitter::code_t)REG_P0;
    assert((ureg >= 8) && (ureg <= 15));
    return (ureg - 8 ) << 0;
}
```

Looking at the autogenerated code, `ptrue` is the only other instruction that requires `insEncodeReg_P_2_to_0()`, and that uses it in the same way. So it looks like it's safe to change.

amanasifkhalid committed:
Fix insEncodeReg_P_2_to_0

amanasifkhalid commented:
@a74nh thank you for the fix! Changing `insEncodeReg_P_2_to_0` fixed the diffs locally.

I'm trying to determine the best way to differentiate these instruction formats in `emitIns_R_R_R` without introducing new `insOpts`, as we only give `_idInsOpt` 6 bits (so 64 possible values) on ARM64, and increasing this allotment doesn't seem trivial. Would we be okay with splitting up these instruction definitions in `instrsarm64sve.h`? For example, `INS_sve_whilege` becomes `INS_sve_whilege1`, `INS_sve_whilege2`, and `INS_sve_whilege3` (or maybe something more descriptive than just a numeric suffix), so now we have different switch cases in `emitIns_R_R_R` to handle the three different formats. I think we'd still need to differentiate the `vl` specifier for `IF_SVE_DY_3A` by the `insOpts` passed to `emitIns_R_R_R`, but this will at least reduce the number of new `insOpts` needed.

Another option is to not split up the instructions, but to (mis-)use existing `insOpts` values to differentiate the formats, but this would probably be confusing to the reader. Maybe we could hide this usage behind helper methods (like `getInsOptsScalableWithPredicatePair(insOpts opt)`, which takes in e.g. `INS_OPTS_SCALABLE_B_WITH_SCALAR` and returns `INS_OPTS_SCALABLE_B_WITH_PREDICATE_MERGE`) to improve readability?

a74nh commented:
> I'm trying to determine the best way to differentiate these instruction formats in `emitIns_R_R_R` without introducing new `insOpts`, as we only give `_idInsOpt` 6 bits (so 64 possible values) on ARM64, and increasing this allotment doesn't seem trivial.

I hit the same issue with a previous PR, but managed to reduce the insopts value for that one. I was a little concerned we'd hit it again.

> Would we be okay with splitting up these instruction definitions in `instrsarm64sve.h`? For example, `INS_sve_whilege` becomes `INS_sve_whilege1`, `INS_sve_whilege2`, and `INS_sve_whilege3` (or maybe something more descriptive than just a numeric suffix), so now we have different switch cases in `emitIns_R_R_R` to handle the three different formats. I think we'd still need to differentiate the `vl` specifier for `IF_SVE_DY_3A` by the `insOpts` passed to `emitIns_R_R_R`, but this will at least reduce the number of new `insOpts` needed.

Ideally I'd rather avoid this. But we might have to eventually.

> Another option is to not split up the instructions, but to (mis-)use existing `insOpts` values to differentiate the formats, but this would probably be confusing to the reader. Maybe we could hide this usage behind helper methods (like `getInsOptsScalableWithPredicatePair(insOpts opt)`, which takes in e.g. `INS_OPTS_SCALABLE_B_WITH_SCALAR` and returns `INS_OPTS_SCALABLE_B_WITH_PREDICATE_MERGE`) to improve readability?

Reducing `insOpts` seems the better way.

One way would be to turn it from an enum into flags:
```
    INS_OPTS_SCALABLE_B = 0x1
    INS_OPTS_SCALABLE_H = 0x2
    INS_OPTS_SCALABLE_S = 0x4
    INS_OPTS_SCALABLE_D = 0x8
    INS_OPTS_SCALABLE_WIDE = 0x10
    INS_OPTS_SCALABLE_WITH_SIMD_VECTOR =0x20
    INS_OPTS_SCALABLE_WITH_SIMD_SCALAR = 0x40
    INS_OPTS_SCALABLE_WITH_SCALAR = 0x80
```
But that's a lot of refactoring code and we might still run out. So, probably not.

There should be no overlap the non-SVE and SVE values. So we should simply be able to restart the enum at 1:
```
INS_OPTS_SCALABLE_B = 0x1
    INS_OPTS_SCALABLE_H,
    INS_OPTS_SCALABLE_S,
    INS_OPTS_SCALABLE_D,
    INS_OPTS_SCALABLE_WIDE_B = 0x5
    INS_OPTS_SCALABLE_WIDE_H,
    INS_OPTS_SCALABLE_WIDE_S,
 etc
```
So, `INS_OPTS_PRE_INDEX` and `INS_OPTS_SCALABLE_B` are both 1. No other code then needs to change.

amanasifkhalid committed:
Reset insOpts enum value for SVE formats

amanasifkhalid commented:
> There should be no overlap the non-SVE and SVE values. So we should simply be able to restart the enum at 1:

Good point, thank you for the suggestion! I had to refactor `emitDispArrangement` to get this to build, but everything seems to be working now. Here's the (sanitized) JitDisasm:

```
whilege p0.b, w0, w1
whilege p1.b, x2, x3
whilegt p2.b, w4, w5
whilegt p3.b, x6, x7
whilehi p4.h, w8, w9
whilehi p5.h, x10, x11
whilehs p6.h, w12, w13
whilehs p7.h, x14, x15
whilele p8.s, w0, w1
whilele p9.s, x2, x3
whilelo p10.s, w4, w5
whilelo p11.s, x6, x7
whilels p12.d, w8, w9
whilels p13.d, x10, x11
whilelt p14.d, w12, w13
whilelt p15.d, x14, x15
whilerw p0.b, x0, x1
whilerw p1.h, x2, x3
whilerw p2.s, x4, x5
whilerw p3.d, x6, x7
whilewr p4.b, x8, x9
whilewr p5.h, x10, x11
whilewr p6.s, x12, x13
whilewr p7.d, x14, x15
whilege { p0.b, p1.b }, x0, x1
whilegt { p2.b, p3.b }, x2, x3
whilehi { p4.h, p5.h }, x4, x5
whilehs { p6.h, p7.h }, x6, x7
whilele { p8.s, p9.s }, x8, x9
whilelo { p10.s, p11.s }, x10, x11
whilels { p12.d, p13.d }, x12, x13
whilelt { p14.d, p15.d }, x14, x15
whilege p8.b, x0, x1, vlx2
whilege p9.b, x2, x3, vlx4
whilegt p10.h, x4, x5, vlx2
whilegt p11.h, x6, x7, vlx4
whilehi p12.s, x8, x9, vlx2
whilehi p13.s, x10, x11, vlx4
whilehs p14.d, x12, x13, vlx2
whilehs p15.d, x14, x15, vlx4
whilele p8.b, x0, x1, vlx2
whilele p9.b, x2, x3, vlx4
whilelo p10.h, x4, x5, vlx2
whilelo p11.h, x6, x7, vlx4
whilels p12.d, x8, x9, vlx2
whilels p13.d, x10, x11, vlx4
whilelt p14.b, x12, x13, vlx2
whilelt p15.b, x14, x15, vlx4
```

and the (sanitized) cstool output:

```
whilege p0.b, w0, w1
whilege p1.b, x2, x3
whilegt p2.b, w4, w5
whilegt p3.b, x6, x7
whilehi p4.h, w8, w9
whilehi p5.h, x10, x11
whilehs p6.h, w12, w13
whilehs p7.h, x14, x15
whilele p8.s, w0, w1
whilele p9.s, x2, x3
whilelo p10.s, w4, w5
whilelo p11.s, x6, x7
whilels p12.d, w8, w9
whilels p13.d, x10, x11
whilelt p14.d, w12, w13
whilelt p15.d, x14, x15
whilerw p0.b, x0, x1
whilerw p1.h, x2, x3
whilerw p2.s, x4, x5
whilerw p3.d, x6, x7
whilewr p4.b, x8, x9
whilewr p5.h, x10, x11
whilewr p6.s, x12, x13
whilewr p7.d, x14, x15
whilege { p0.b, p1.b }, x0, x1
whilegt { p2.b, p3.b }, x2, x3
whilehi { p4.h, p5.h }, x4, x5
whilehs { p6.h, p7.h }, x6, x7
whilele { p8.s, p9.s }, x8, x9
whilelo { p10.s, p11.s }, x10, x11
whilels { p12.d, p13.d }, x12, x13
whilelt { p14.d, p15.d }, x14, x15
whilege pn8.b, x0, x1, vlx2
whilege pn9.b, x2, x3, vlx4
whilegt pn0xA.h, x4, x5, vlx2
whilegt pn0xB.h, x6, x7, vlx4
whilehi pn0xC.s, x8, x9, vlx2
whilehi pn0xD.s, x10, x11, vlx4
whilehs pn0xE.d, x12, x13, vlx2
whilehs pn0xF.d, x14, x15, vlx4
whilele pn8.b, x0, x1, vlx2
whilele pn9.b, x2, x3, vlx4
whilelo pn0xA.h, x4, x5, vlx2
whilelo pn0xB.h, x6, x7, vlx4
whilels pn0xC.d, x8, x9, vlx2
whilels pn0xD.d, x10, x11, vlx4
whilelt pn0xE.b, x12, x13, vlx2
whilelt pn0xF.b, x14, x15, vlx4
```

The only diffs I see locally are the different predicate register formats at the end (e.g. `p10.h` vs `p0xA.h`)

a74nh reviewed (approved):
LGTM now. Thanks!

(note: I'm away until new year now).

amanasifkhalid commented:
Thank you for the review, and enjoy your holiday!

kunalspathak commented on src/coreclr/jit/emitarm64.h:
```diff
@@ -789,12 +795,17 @@ inline static bool isFloatReg(regNumber reg)
+
+inline static bool isHighPredicateRegister(regNumber reg)
+{
+    return (reg > REG_PREDICATE_LOW_LAST) && (reg <= REG_PREDICATE_LAST);
```
probably good to introduce `REG_PREDICATE_HIGH_FIRST` and `REG_PREDICATE_HIGH_LAST` and use it here. It is more explicit that way.

kunalspathak replied:
and a `static_assert_no_msg(REG_PREDICATE_HIGH_LAST == REG_PREDICATE_LAST)`.

kunalspathak commented on src/coreclr/jit/emitarm64.cpp:
```diff
@@ -15658,6 +15844,44 @@ void emitter::emitDispArrangement(insOpts opt)
             str = "8b";
             break;
         case INS_OPTS_16B:
+            str = "16b";
```
any reason why these were added in `emitDispArrangement`?

amanasifkhalid replied:
Those cases were already in `emitDispArrangement`, I think the diff just makes it look like they were added in. I moved all the SVE-specific cases to `emitDispSveArrangement`, and left the remaining cases in `emitDispArrangement` as-is.

kunalspathak commented on src/coreclr/jit/emitarm64.cpp:
```diff
@@ -15668,10 +15892,6 @@ void emitter::emitDispArrangement(insOpts opt)
         case INS_OPTS_SCALABLE_B_WITH_PREDICATE_MERGE:
             str = "b";
             break;
-        case INS_OPTS_4H:
```
does removing these have any implications on existing instructions that we added so far? Can you try to compare the output of all instructions jitdisasm vs. cstool with this change?

amanasifkhalid replied:
> Can you try to compare the output of all instructions jitdisasm vs. cstool with this change?

Sure thing. I'll post a comment with diffable outputs from both.

kunalspathak reviewed (changes requested):
Added few questions and suggestions.

amanasifkhalid commented:
Here are ugly but diffable JitDisasm/cstool outputs for the SVE instructions we have so far. I don't see any diffs locally, except for the different predicate register encodings we noted above.

JitDisasm:
```
andz0.b,p1/m,z0.b,z2.b
bicz3.h,p4/m,z3.h,z5.h
eorz14.s,p5/m,z14.s,z16.s
orrz29.d,p7/m,z29.d,z31.d
addz5.b,p6/m,z5.b,z7.b
subz15.h,p7/m,z15.h,z29.h
subrz2.s,p0/m,z2.s,z13.s
sdivz3.s,p2/m,z3.s,z9.s
sdivrz31.d,p3/m,z31.d,z29.d
udivz1.s,p0/m,z1.s,z0.s
udivrz13.d,p7/m,z13.d,z15.d
smaxz24.b,p0/m,z24.b,z2.b
sminz9.h,p1/m,z9.h,z27.h
sabdz5.b,p2/m,z5.b,z6.b
uabdz23.s,p3/m,z23.s,z9.s
umaxz15.s,p4/m,z15.s,z2.s
uminz12.d,p7/m,z12.d,z0.d
mulz5.d,p1/m,z5.d,z3.d
smulhz17.s,p5/m,z17.s,z5.s
umulhz12.b,p2/m,z12.b,z24.b
movprfxz13.b,p0/z,z31.b
movprfxz14.h,p1/z,z0.h
movprfxz13.s,p2/z,z1.s
movprfxz15.d,p3/z,z2.d
movprfxz16.b,p4/m,z3.b
movprfxz17.h,p5/m,z12.h
movprfxz0.s,p6/m,z13.s
movprfxz31.d,p7/m,z22.d
asrz0.b,p0/m,z0.b,#1
asrz1.b,p1/m,z1.b,#2
asrz2.b,p2/m,z2.b,#3
asrz3.b,p3/m,z3.b,#7
asrz4.b,p7/m,z4.b,#8
asrdz5.h,p0/m,z5.h,#1
asrdz6.h,p4/m,z6.h,#8
asrdz7.h,p7/m,z7.h,#15
asrdz31.h,p6/m,z31.h,#16
lsrz8.s,p5/m,z8.s,#1
lsrz9.s,p6/m,z9.s,#5
lsrz31.s,p7/m,z31.s,#17
lsrz0.s,p0/m,z0.s,#31
lsrz5.s,p1/m,z5.s,#32
srshrz6.d,p3/m,z6.d,#1
srshrz7.d,p4/m,z7.d,#3
srshrz21.d,p5/m,z21.d,#21
srshrz22.d,p6/m,z22.d,#63
urshrz31.d,p7/m,z31.d,#64
lslz1.b,p0/m,z1.b,#0
lslz3.b,p1/m,z3.b,#1
lslz5.b,p2/m,z5.b,#7
sqshlz7.h,p3/m,z7.h,#0
sqshlz9.h,p4/m,z9.h,#7
sqshlz11.h,p0/m,z11.h,#9
sqshlz13.h,p7/m,z13.h,#15
sqshluz14.s,p6/m,z14.s,#0
sqshluz15.s,p0/m,z15.s,#12
sqshluz16.s,p1/m,z16.s,#15
sqshluz0.s,p2/m,z0.s,#31
uqshlz17.d,p7/m,z17.d,#0
uqshlz18.d,p0/m,z18.d,#18
uqshlz19.d,p3/m,z19.d,#32
uqshlz0.d,p2/m,z0.d,#63
asrz5.s,p0/m,z5.s,z21.s
asrrz1.b,p7/m,z1.b,z20.b
lslz0.h,p2/m,z0.h,z0.h
lslrz27.d,p6/m,z27.d,z31.d
lsrz5.b,p5/m,z5.b,z6.b
lsrrz15.s,p4/m,z15.s,z17.s
asrz4.b,p3/m,z4.b,z24.d
lslz19.h,p7/m,z19.h,z3.d
lsrz0.s,p0/m,z0.s,z0.d
clastaz31.b,p7,z31.b,z31.b
clastbz30.d,p6,z30.d,z30.d
clastah12,p1,h12,z15.h
clastbs13,p2,s13,z16.s
clastbd14,p0,d14,z17.d
clastaw0,p0,w0,z0.b
clastaw1,p2,w1,z3.h
clastbw23,p5,w23,z12.s
clastbx3,p6,x3,z9.d
cmpeqp15.b,p0/z,z0.b,z10.b
cmpgep14.h,p1/z,z10.h,z23.h
cmpgtp13.s,p2/z,z20.s,z24.s
cmphip12.d,p3/z,z30.d,z25.d
cmphsp11.b,p4/z,z1.b,z26.b
cmpgep10.h,p5/z,z27.h,z11.h
cmphip11.s,p6/z,z28.s,z12.s
cmphsp0.d,p7/z,z29.d,z13.d
cmpgtp2.b,p0/z,z30.b,z14.b
cmpnep3.h,p1/z,z15.h,z20.h
shaddz15.b,p0/m,z15.b,z10.b
shsubz16.h,p1/m,z16.h,z11.h
shsubrz17.s,p2/m,z17.s,z12.s
srhaddz18.d,p3/m,z18.d,z13.d
uhaddz19.b,p4/m,z19.b,z14.b
uhsubz20.h,p5/m,z20.h,z15.h
uhsubrz21.s,p6/m,z21.s,z16.s
urhaddz22.d,p7/m,z22.d,z17.d
addpz23.b,p6/m,z23.b,z18.b
smaxpz24.h,p5/m,z24.h,z19.h
sminpz25.s,p4/m,z25.s,z20.s
umaxpz26.d,p3/m,z26.d,z21.d
uminpz27.b,p2/m,z27.b,z22.b
sqaddz28.b,p1/m,z28.b,z23.b
sqsubz29.h,p0/m,z29.h,z24.h
sqsubrz30.h,p1/m,z30.h,z25.h
suqaddz31.b,p2/m,z31.b,z26.b
uqaddz0.s,p3/m,z0.s,z27.s
uqsubz1.d,p4/m,z1.d,z28.d
uqsubrz2.b,p5/m,z2.b,z29.b
usqaddz3.b,p6/m,z3.b,z30.b
sqrshlz4.b,p7/m,z4.b,z31.b
sqrshlrz5.h,p0/m,z5.h,z30.h
sqshlz6.s,p1/m,z6.s,z29.s
sqshlrz7.d,p2/m,z7.d,z28.d
srshlz8.b,p3/m,z8.b,z27.b
srshlrz9.h,p4/m,z9.h,z26.h
uqrshlz10.s,p5/m,z10.s,z25.s
uqrshlrz11.d,p6/m,z11.d,z24.d
uqshlz12.b,p7/m,z12.b,z23.b
uqshlrz13.h,p0/m,z13.h,z22.h
urshlz14.s,p1/m,z14.s,z21.s
urshlrz15.d,p2/m,z15.d,z20.d
faddpz16.h,p3/m,z16.h,z19.h
fmaxnmpz17.s,p4/m,z17.s,z18.s
fmaxpz18.d,p5/m,z18.d,z17.d
fminnmpz19.s,p6/m,z19.s,z16.s
fminpz20.h,p7/m,z20.h,z15.h
faddah21,p6,h21,z14.h
faddas22,p5,s22,z13.s
faddad23,p4,d23,z12.d
fabdz24.h,p3/m,z24.h,z11.h
faddz25.s,p2/m,z25.s,z10.s
fdivz28.s,p0/m,z28.s,z7.s
fdivrz29.d,p1/m,z29.d,z6.d
fmaxz30.h,p2/m,z30.h,z5.h
fmaxnmz31.s,p3/m,z31.s,z4.s
fminz0.d,p4/m,z0.d,z3.d
fminnmz1.h,p5/m,z1.h,z2.h
fmulz2.s,p6/m,z2.s,z1.s
fmulxz3.d,p7/m,z3.d,z0.d
fscalez4.h,p6/m,z4.h,z31.h
fsubz5.s,p5/m,z5.s,z30.s
fsubrz6.d,p4/m,z6.d,z29.d
andvb0,p0,z0.b
eorvh1,p1,z1.h
orvs2,p2,z2.s
orvd3,p3,z3.d
saddvd1,p4,z2.b
saddvd2,p5,z3.h
uaddvd3,p6,z4.s
smaxvd15,p7,z4.d
sminvs16,p6,z14.s
umaxvh17,p5,z24.h
uminvb18,p4,z31.b
clsz31.b,p0/m,z0.b
clzz30.h,p1/m,z1.h
cnotz29.s,p2/m,z2.s
cntz28.d,p3/m,z3.d
fabsz27.h,p4/m,z4.h
fnegz26.s,p5/m,z5.s
notz25.b,p6/m,z6.b
absz24.b,p7/m,z7.b
negz23.s,p0/m,z8.s
sxtbz22.h,p1/m,z9.h
sxtbz22.s,p1/m,z9.s
sxtbz22.d,p1/m,z9.d
sxthz21.s,p2/m,z10.s
sxthz21.d,p2/m,z10.d
sxtwz20.d,p3/m,z11.d
uxtbz19.h,p4/m,z12.h
uxtbz19.s,p4/m,z12.s
uxtbz19.d,p4/m,z12.d
uxthz18.s,p5/m,z13.s
uxthz18.d,p5/m,z13.d
uxtwz17.d,p6/m,z14.d
mlaz0.b,p0/m,z0.b,z19.b
mlsz2.h,p1/m,z31.h,z31.h
madz4.s,p2/m,z23.s,z12.s
msbz8.d,p7/m,z12.d,z22.d
compactz16.s,p7,z13.s
compactz15.d,p0,z12.d
movz14.b,p1/m,b11
movz13.s,p2/m,s10
movz12.h,p3/m,h9
movz11.d,p4/m,d8
movz10.d,p5/m,sp
movz9.h,p6/m,w30
movz8.s,p7/m,w29
movz7.b,p0/m,w28
lastab6,p1,z27.b
lastah5,p2,z26.h
lastbs4,p3,z25.s
lastbd3,p4,z24.d
lastaw1,p5,z23.b
lastaw0,p6,z22.s
lastbw30,p7,z21.h
lastbfp,p0,z20.d
rbitz28.h,p1/m,z19.h
rbitz28.b,p1/m,z19.b
rbitz28.s,p1/m,z19.s
rbitz28.d,p1/m,z19.d
revbz27.h,p2/m,z18.h
revbz27.s,p2/m,z18.s
revbz27.d,p2/m,z18.d
revhz26.s,p3/m,z17.s
revhz26.d,p3/m,z17.d
revwz25.d,p4/m,z16.d
sadalpz26.h,p3/m,z8.b
uadalpz27.s,p2/m,z9.h
uadalpz28.d,p0/m,z31.s
sqabsz29.b,p7/m,z0.b
sqnegz31.h,p6/m,z1.h
sqnegz0.s,p5/m,z2.s
sqnegz1.d,p4/m,z3.d
urecpez2.s,p3/m,z4.s
ursqrtez3.s,p0/m,z5.s
sqrshrnz0.h,{z0.s,z1.s},#5
sqrshrunz0.h,{z0.s,z1.s},#5
uqrshrnz0.h,{z0.s,z1.s},#5
sqrshrnz0.h,{z2.s,z3.s},#16
sqrshrunz0.h,{z4.s,z5.s},#7
uqrshrnz0.h,{z6.s,z7.s},#1
sqrshrnz30.h,{z16.s,z17.s},#16
sqrshrunz16.h,{z8.s,z9.s},#7
uqrshrnz15.h,{z12.s,z13.s},#1
decpx0,p0.b
decpx1,p1.h
decpx2,p2.s
decpx3,p3.d
incpx4,p4.b
incpx5,p5.h
incpx6,p6.s
incpx7,p7.d
decpz0.h,p0.h
decpz1.s,p1.s
decpz2.d,p2.d
incpz3.h,p3.h
incpz4.s,p4.s
incpz5.d,p5.d
sqdecpx0,p0.b,w0
sqdecpx1,p1.h,w1
sqdecpx2,p2.s,w2
sqdecpx3,p3.d,w3
sqdecpx4,p4.b
sqdecpx5,p5.h
sqdecpx6,p6.s
sqdecpx7,p7.d
sqincpx0,p0.h,w0
sqincpx1,p1.s,w1
sqincpx2,p2.b,w2
sqincpx3,p3.d,w3
sqincpx4,p4.b
sqincpx5,p5.h
sqincpx6,p6.s
sqincpx7,p7.d
uqdecpw0,p0.b
uqdecpw1,p1.h
uqdecpw2,p2.s
uqdecpw3,p3.d
uqdecpx4,p4.b
uqdecpx5,p5.h
uqdecpx6,p6.s
uqdecpx7,p7.d
uqincpw0,p0.b
uqincpw1,p1.h
uqincpw2,p2.s
uqincpw3,p3.d
uqincpx4,p4.b
uqincpx5,p5.h
uqincpx6,p6.s
uqincpx7,p7.d
sqdecpz0.h,p0.h
sqdecpz1.s,p1.s
sqdecpz2.d,p2.d
sqincpz3.h,p3.h
sqincpz4.s,p4.s
sqincpz5.d,p5.d
uqdecpz6.h,p6.h
uqdecpz7.s,p7.s
uqdecpz8.d,p0.d
uqincpz9.h,p1.h
uqincpz10.s,p2.s
uqincpz11.d,p3.d
setffr
wrffrp0.b
ctermeqx0,x1
ctermeqw2,w3
ctermnex4,x5
ctermnew6,w7
sqxtnbz0.b,z5.h
sqxtntz3.b,z7.h
sqxtunbz0.b,z8.h
sqxtuntz5.b,z3.h
uqxtnbz0.b,z7.h
uqxtntz0.b,z1.h
sqxtnbz0.h,z0.s
sqxtntz9.h,z9.s
sqxtunbz0.h,z6.s
sqxtuntz0.h,z4.s
uqxtnbz6.h,z2.s
uqxtntz0.h,z1.s
sqxtnbz7.s,z7.d
sqxtntz0.s,z8.d
sqxtunbz0.s,z9.d
sqxtuntz0.s,z4.d
uqxtnbz3.s,z9.d
uqxtntz0.s,z3.d
aesdz0.b,z0.b,z0.b
aesez1.b,z1.b,z2.b
sm4ez3.s,z3.s,z5.s
aesimcz0.b,z0.b
aesmcz5.b,z5.b
faddvh21,p7,z7.h
fmaxnmvh22,p6,z6.h
fmaxvs23,p5,z5.s
fminnmvd24,p4,z4.d
fminvs25,p3,z3.s
frintaz26.h,p7/m,z2.h
frintiz27.s,p6/m,z1.s
frintmz28.d,p5/m,z0.d
frintnz29.h,p4/m,z10.h
frintpz30.s,p3/m,z11.s
frintxz31.d,p2/m,z12.d
frintzz0.h,p0/m,z13.h
frecpxz5.h,p5/m,z5.h
fsqrtz6.s,p6/m,z6.s
whilegep0.b,w0,w1
whilegep1.b,x2,x3
whilegtp2.b,w4,w5
whilegtp3.b,x6,x7
whilehip4.h,w8,w9
whilehip5.h,x10,x11
whilehsp6.h,w12,w13
whilehsp7.h,x14,x15
whilelep8.s,w0,w1
whilelep9.s,x2,x3
whilelop10.s,w4,w5
whilelop11.s,x6,x7
whilelsp12.d,w8,w9
whilelsp13.d,x10,x11
whileltp14.d,w12,w13
whileltp15.d,x14,x15
whilerwp0.b,x0,x1
whilerwp1.h,x2,x3
whilerwp2.s,x4,x5
whilerwp3.d,x6,x7
whilewrp4.b,x8,x9
whilewrp5.h,x10,x11
whilewrp6.s,x12,x13
whilewrp7.d,x14,x15
whilege{p0.b,p1.b},x0,x1
whilegt{p2.b,p3.b},x2,x3
whilehi{p4.h,p5.h},x4,x5
whilehs{p6.h,p7.h},x6,x7
whilele{p8.s,p9.s},x8,x9
whilelo{p10.s,p11.s},x10,x11
whilels{p12.d,p13.d},x12,x13
whilelt{p14.d,p15.d},x14,x15
whilegep8.b,x0,x1,vlx2
whilegep9.b,x2,x3,vlx4
whilegtp10.h,x4,x5,vlx2
whilegtp11.h,x6,x7,vlx4
whilehip12.s,x8,x9,vlx2
whilehip13.s,x10,x11,vlx4
whilehsp14.d,x12,x13,vlx2
whilehsp15.d,x14,x15,vlx4
whilelep8.b,x0,x1,vlx2
whilelep9.b,x2,x3,vlx4
whilelop10.h,x4,x5,vlx2
whilelop11.h,x6,x7,vlx4
whilelsp12.d,x8,x9,vlx2
whilelsp13.d,x10,x11,vlx4
whileltp14.b,x12,x13,vlx2
whileltp15.b,x14,x15,vlx4
```

cstool:

```
andz0.b,p1/m,z0.b,z2.b
bicz3.h,p4/m,z3.h,z5.h
eorz14.s,p5/m,z14.s,z16.s
orrz29.d,p7/m,z29.d,z31.d
addz5.b,p6/m,z5.b,z7.b
subz15.h,p7/m,z15.h,z29.h
subrz2.s,p0/m,z2.s,z13.s
sdivz3.s,p2/m,z3.s,z9.s
sdivrz31.d,p3/m,z31.d,z29.d
udivz1.s,p0/m,z1.s,z0.s
udivrz13.d,p7/m,z13.d,z15.d
smaxz24.b,p0/m,z24.b,z2.b
sminz9.h,p1/m,z9.h,z27.h
sabdz5.b,p2/m,z5.b,z6.b
uabdz23.s,p3/m,z23.s,z9.s
umaxz15.s,p4/m,z15.s,z2.s
uminz12.d,p7/m,z12.d,z0.d
mulz5.d,p1/m,z5.d,z3.d
smulhz17.s,p5/m,z17.s,z5.s
umulhz12.b,p2/m,z12.b,z24.b
movprfxz13.b,p0/z,z31.b
movprfxz14.h,p1/z,z0.h
movprfxz13.s,p2/z,z1.s
movprfxz15.d,p3/z,z2.d
movprfxz16.b,p4/m,z3.b
movprfxz17.h,p5/m,z12.h
movprfxz0.s,p6/m,z13.s
movprfxz31.d,p7/m,z22.d
asrz0.b,p0/m,z0.b,#1
asrz1.b,p1/m,z1.b,#2
asrz2.b,p2/m,z2.b,#3
asrz3.b,p3/m,z3.b,#7
asrz4.b,p7/m,z4.b,#8
asrdz5.h,p0/m,z5.h,#1
asrdz6.h,p4/m,z6.h,#8
asrdz7.h,p7/m,z7.h,#15
asrdz31.h,p6/m,z31.h,#16
lsrz8.s,p5/m,z8.s,#1
lsrz9.s,p6/m,z9.s,#5
lsrz31.s,p7/m,z31.s,#17
lsrz0.s,p0/m,z0.s,#31
lsrz5.s,p1/m,z5.s,#32
srshrz6.d,p3/m,z6.d,#1
srshrz7.d,p4/m,z7.d,#3
srshrz21.d,p5/m,z21.d,#21
srshrz22.d,p6/m,z22.d,#63
urshrz31.d,p7/m,z31.d,#64
lslz1.b,p0/m,z1.b,#0
lslz3.b,p1/m,z3.b,#1
lslz5.b,p2/m,z5.b,#7
sqshlz7.h,p3/m,z7.h,#0
sqshlz9.h,p4/m,z9.h,#7
sqshlz11.h,p0/m,z11.h,#9
sqshlz13.h,p7/m,z13.h,#15
sqshluz14.s,p6/m,z14.s,#0
sqshluz15.s,p0/m,z15.s,#12
sqshluz16.s,p1/m,z16.s,#15
sqshluz0.s,p2/m,z0.s,#31
uqshlz17.d,p7/m,z17.d,#0
uqshlz18.d,p0/m,z18.d,#18
uqshlz19.d,p3/m,z19.d,#32
uqshlz0.d,p2/m,z0.d,#63
asrz5.s,p0/m,z5.s,z21.s
asrrz1.b,p7/m,z1.b,z20.b
lslz0.h,p2/m,z0.h,z0.h
lslrz27.d,p6/m,z27.d,z31.d
lsrz5.b,p5/m,z5.b,z6.b
lsrrz15.s,p4/m,z15.s,z17.s
asrz4.b,p3/m,z4.b,z24.d
lslz19.h,p7/m,z19.h,z3.d
lsrz0.s,p0/m,z0.s,z0.d
clastaz31.b,p7,z31.b,z31.b
clastbz30.d,p6,z30.d,z30.d
clastah12,p1,h12,z15.h
clastbs13,p2,s13,z16.s
clastbd14,p0,d14,z17.d
clastaw0,p0,w0,z0.b
clastaw1,p2,w1,z3.h
clastbw23,p5,w23,z12.s
clastbx3,p6,x3,z9.d
cmpeqp15.b,p0/z,z0.b,z10.b
cmpgep14.h,p1/z,z10.h,z23.h
cmpgtp13.s,p2/z,z20.s,z24.s
cmphip12.d,p3/z,z30.d,z25.d
cmphsp11.b,p4/z,z1.b,z26.b
cmpgep10.h,p5/z,z27.h,z11.h
cmphip11.s,p6/z,z28.s,z12.s
cmphsp0.d,p7/z,z29.d,z13.d
cmpgtp2.b,p0/z,z30.b,z14.b
cmpnep3.h,p1/z,z15.h,z20.h
shaddz15.b,p0/m,z15.b,z10.b
shsubz16.h,p1/m,z16.h,z11.h
shsubrz17.s,p2/m,z17.s,z12.s
srhaddz18.d,p3/m,z18.d,z13.d
uhaddz19.b,p4/m,z19.b,z14.b
uhsubz20.h,p5/m,z20.h,z15.h
uhsubrz21.s,p6/m,z21.s,z16.s
urhaddz22.d,p7/m,z22.d,z17.d
addpz23.b,p6/m,z23.b,z18.b
smaxpz24.h,p5/m,z24.h,z19.h
sminpz25.s,p4/m,z25.s,z20.s
umaxpz26.d,p3/m,z26.d,z21.d
uminpz27.b,p2/m,z27.b,z22.b
sqaddz28.b,p1/m,z28.b,z23.b
sqsubz29.h,p0/m,z29.h,z24.h
sqsubrz30.h,p1/m,z30.h,z25.h
suqaddz31.b,p2/m,z31.b,z26.b
uqaddz0.s,p3/m,z0.s,z27.s
uqsubz1.d,p4/m,z1.d,z28.d
uqsubrz2.b,p5/m,z2.b,z29.b
usqaddz3.b,p6/m,z3.b,z30.b
sqrshlz4.b,p7/m,z4.b,z31.b
sqrshlrz5.h,p0/m,z5.h,z30.h
sqshlz6.s,p1/m,z6.s,z29.s
sqshlrz7.d,p2/m,z7.d,z28.d
srshlz8.b,p3/m,z8.b,z27.b
srshlrz9.h,p4/m,z9.h,z26.h
uqrshlz10.s,p5/m,z10.s,z25.s
uqrshlrz11.d,p6/m,z11.d,z24.d
uqshlz12.b,p7/m,z12.b,z23.b
uqshlrz13.h,p0/m,z13.h,z22.h
urshlz14.s,p1/m,z14.s,z21.s
urshlrz15.d,p2/m,z15.d,z20.d
faddpz16.h,p3/m,z16.h,z19.h
fmaxnmpz17.s,p4/m,z17.s,z18.s
fmaxpz18.d,p5/m,z18.d,z17.d
fminnmpz19.s,p6/m,z19.s,z16.s
fminpz20.h,p7/m,z20.h,z15.h
faddah21,p6,h21,z14.h
faddas22,p5,s22,z13.s
faddad23,p4,d23,z12.d
fabdz24.h,p3/m,z24.h,z11.h
faddz25.s,p2/m,z25.s,z10.s
fdivz28.s,p0/m,z28.s,z7.s
fdivrz29.d,p1/m,z29.d,z6.d
fmaxz30.h,p2/m,z30.h,z5.h
fmaxnmz31.s,p3/m,z31.s,z4.s
fminz0.d,p4/m,z0.d,z3.d
fminnmz1.h,p5/m,z1.h,z2.h
fmulz2.s,p6/m,z2.s,z1.s
fmulxz3.d,p7/m,z3.d,z0.d
fscalez4.h,p6/m,z4.h,z31.h
fsubz5.s,p5/m,z5.s,z30.s
fsubrz6.d,p4/m,z6.d,z29.d
andvb0,p0,z0.b
eorvh1,p1,z1.h
orvs2,p2,z2.s
orvd3,p3,z3.d
saddvd1,p4,z2.b
saddvd2,p5,z3.h
uaddvd3,p6,z4.s
smaxvd15,p7,z4.d
sminvs16,p6,z14.s
umaxvh17,p5,z24.h
uminvb18,p4,z31.b
clsz31.b,p0/m,z0.b
clzz30.h,p1/m,z1.h
cnotz29.s,p2/m,z2.s
cntz28.d,p3/m,z3.d
fabsz27.h,p4/m,z4.h
fnegz26.s,p5/m,z5.s
notz25.b,p6/m,z6.b
absz24.b,p7/m,z7.b
negz23.s,p0/m,z8.s
sxtbz22.h,p1/m,z9.h
sxtbz22.s,p1/m,z9.s
sxtbz22.d,p1/m,z9.d
sxthz21.s,p2/m,z10.s
sxthz21.d,p2/m,z10.d
sxtwz20.d,p3/m,z11.d
uxtbz19.h,p4/m,z12.h
uxtbz19.s,p4/m,z12.s
uxtbz19.d,p4/m,z12.d
uxthz18.s,p5/m,z13.s
uxthz18.d,p5/m,z13.d
uxtwz17.d,p6/m,z14.d
mlaz0.b,p0/m,z0.b,z19.b
mlsz2.h,p1/m,z31.h,z31.h
madz4.s,p2/m,z23.s,z12.s
msbz8.d,p7/m,z12.d,z22.d
compactz16.s,p7,z13.s
compactz15.d,p0,z12.d
movz14.b,p1/m,b11
movz13.s,p2/m,s10
movz12.h,p3/m,h9
movz11.d,p4/m,d8
movz10.d,p5/m,sp
movz9.h,p6/m,w30
movz8.s,p7/m,w29
movz7.b,p0/m,w28
lastab6,p1,z27.b
lastah5,p2,z26.h
lastbs4,p3,z25.s
lastbd3,p4,z24.d
lastaw1,p5,z23.b
lastaw0,p6,z22.s
lastbw30,p7,z21.h
lastbfp,p0,z20.d
rbitz28.h,p1/m,z19.h
rbitz28.b,p1/m,z19.b
rbitz28.s,p1/m,z19.s
rbitz28.d,p1/m,z19.d
revbz27.h,p2/m,z18.h
revbz27.s,p2/m,z18.s
revbz27.d,p2/m,z18.d
revhz26.s,p3/m,z17.s
revhz26.d,p3/m,z17.d
revwz25.d,p4/m,z16.d
sadalpz26.h,p3/m,z8.b
uadalpz27.s,p2/m,z9.h
uadalpz28.d,p0/m,z31.s
sqabsz29.b,p7/m,z0.b
sqnegz31.h,p6/m,z1.h
sqnegz0.s,p5/m,z2.s
sqnegz1.d,p4/m,z3.d
urecpez2.s,p3/m,z4.s
ursqrtez3.s,p0/m,z5.s
sqrshrnz0.h,{z0.s,z1.s},#5
sqrshrunz0.h,{z0.s,z1.s},#5
uqrshrnz0.h,{z0.s,z1.s},#5
sqrshrnz0.h,{z2.s,z3.s},#16
sqrshrunz0.h,{z4.s,z5.s},#7
uqrshrnz0.h,{z6.s,z7.s},#1
sqrshrnz30.h,{z16.s,z17.s},#16
sqrshrunz16.h,{z8.s,z9.s},#7
uqrshrnz15.h,{z12.s,z13.s},#1
decpx0,p0.b
decpx1,p1.h
decpx2,p2.s
decpx3,p3.d
incpx4,p4.b
incpx5,p5.h
incpx6,p6.s
incpx7,p7.d
decpz0.h,p0.h
decpz1.s,p1.s
decpz2.d,p2.d
incpz3.h,p3.h
incpz4.s,p4.s
incpz5.d,p5.d
sqdecpx0,p0.b,w0
sqdecpx1,p1.h,w1
sqdecpx2,p2.s,w2
sqdecpx3,p3.d,w3
sqdecpx4,p4.b
sqdecpx5,p5.h
sqdecpx6,p6.s
sqdecpx7,p7.d
sqincpx0,p0.h,w0
sqincpx1,p1.s,w1
sqincpx2,p2.b,w2
sqincpx3,p3.d,w3
sqincpx4,p4.b
sqincpx5,p5.h
sqincpx6,p6.s
sqincpx7,p7.d
uqdecpw0,p0.b
uqdecpw1,p1.h
uqdecpw2,p2.s
uqdecpw3,p3.d
uqdecpx4,p4.b
uqdecpx5,p5.h
uqdecpx6,p6.s
uqdecpx7,p7.d
uqincpw0,p0.b
uqincpw1,p1.h
uqincpw2,p2.s
uqincpw3,p3.d
uqincpx4,p4.b
uqincpx5,p5.h
uqincpx6,p6.s
uqincpx7,p7.d
sqdecpz0.h,p0.h
sqdecpz1.s,p1.s
sqdecpz2.d,p2.d
sqincpz3.h,p3.h
sqincpz4.s,p4.s
sqincpz5.d,p5.d
uqdecpz6.h,p6.h
uqdecpz7.s,p7.s
uqdecpz8.d,p0.d
uqincpz9.h,p1.h
uqincpz10.s,p2.s
uqincpz11.d,p3.d
setffr
wrffrp0.b
ctermeqx0,x1
ctermeqw2,w3
ctermnex4,x5
ctermnew6,w7
sqxtnbz0.b,z5.h
sqxtntz3.b,z7.h
sqxtunbz0.b,z8.h
sqxtuntz5.b,z3.h
uqxtnbz0.b,z7.h
uqxtntz0.b,z1.h
sqxtnbz0.h,z0.s
sqxtntz9.h,z9.s
sqxtunbz0.h,z6.s
sqxtuntz0.h,z4.s
uqxtnbz6.h,z2.s
uqxtntz0.h,z1.s
sqxtnbz7.s,z7.d
sqxtntz0.s,z8.d
sqxtunbz0.s,z9.d
sqxtuntz0.s,z4.d
uqxtnbz3.s,z9.d
uqxtntz0.s,z3.d
aesdz0.b,z0.b,z0.b
aesez1.b,z1.b,z2.b
sm4ez3.s,z3.s,z5.s
aesimcz0.b,z0.b
aesmcz5.b,z5.b
faddvh21,p7,z7.h
fmaxnmvh22,p6,z6.h
fmaxvs23,p5,z5.s
fminnmvd24,p4,z4.d
fminvs25,p3,z3.s
frintaz26.h,p7/m,z2.h
frintiz27.s,p6/m,z1.s
frintmz28.d,p5/m,z0.d
frintnz29.h,p4/m,z10.h
frintpz30.s,p3/m,z11.s
frintxz31.d,p2/m,z12.d
frintzz0.h,p0/m,z13.h
frecpxz5.h,p5/m,z5.h
fsqrtz6.s,p6/m,z6.s
whilegep0.b,w0,w1
whilegep1.b,x2,x3
whilegtp2.b,w4,w5
whilegtp3.b,x6,x7
whilehip4.h,w8,w9
whilehip5.h,x10,x11
whilehsp6.h,w12,w13
whilehsp7.h,x14,x15
whilelep8.s,w0,w1
whilelep9.s,x2,x3
whilelop10.s,w4,w5
whilelop11.s,x6,x7
whilelsp12.d,w8,w9
whilelsp13.d,x10,x11
whileltp14.d,w12,w13
whileltp15.d,x14,x15
whilerwp0.b,x0,x1
whilerwp1.h,x2,x3
whilerwp2.s,x4,x5
whilerwp3.d,x6,x7
whilewrp4.b,x8,x9
whilewrp5.h,x10,x11
whilewrp6.s,x12,x13
whilewrp7.d,x14,x15
whilege{p0.b,p1.b},x0,x1
whilegt{p2.b,p3.b},x2,x3
whilehi{p4.h,p5.h},x4,x5
whilehs{p6.h,p7.h},x6,x7
whilele{p8.s,p9.s},x8,x9
whilelo{p10.s,p11.s},x10,x11
whilels{p12.d,p13.d},x12,x13
whilelt{p14.d,p15.d},x14,x15
whilegepn8.b,x0,x1,vlx2
whilegepn9.b,x2,x3,vlx4
whilegtpn0xA.h,x4,x5,vlx2
whilegtpn0xB.h,x6,x7,vlx4
whilehipn0xC.s,x8,x9,vlx2
whilehipn0xD.s,x10,x11,vlx4
whilehspn0xE.d,x12,x13,vlx2
whilehspn0xF.d,x14,x15,vlx4
whilelepn8.b,x0,x1,vlx2
whilelepn9.b,x2,x3,vlx4
whilelopn0xA.h,x4,x5,vlx2
whilelopn0xB.h,x6,x7,vlx4
whilelspn0xC.d,x8,x9,vlx2
whilelspn0xD.d,x10,x11,vlx4
whileltpn0xE.b,x12,x13,vlx2
whileltpn0xF.b,x14,x15,vlx4
```

amanasifkhalid committed:
Feedback

kunalspathak commented:
> I created ugly but diffable JitDisasm/cstool outputs for the SVE instructions we have so far. I don't see any diffs locally, except for the different predicate register encodings we noted above.
> 
> [Gist](https://gist.github.com/amanasifkhalid/3c9516ea14c1c8e991a3af246fd9d6d6)

Thanks for doing this.

kunalspathak commented:
Looks like there are [several test failures](https://dev.azure.com/dnceng-public/public/_build/results?buildId=506757&view=ms.vss-test-web.build-test-results-tab&runId=11822648&resultId=111206&paneView=debug) because of moving around the enum ordering.

```
Assert failure(PID 7640 [0x00001dd8], Thread: 4468 [0x1174]): Assertion failed '!insOptsScalable(opt)' in 'Program:AdvSimd_CompareEqual_Vector64_Byte_Zero(System.Runtime.Intrinsics.Vector64`1[ubyte]):System.Runtime.Intrinsics.Vector64`1[ubyte]' during 'Emit code' (IL size 12; hash 0xb87bf868; FullOpts)

    File: D:\a\_work\1\s\src\coreclr\jit\emitarm64.cpp Line: 15838
    Image: C:\h\w\AC960933\p\corerun.exe
```

amanasifkhalid committed:
Remove asserts

amanasifkhalid commented:
> Looks like there are [several test failures](https://dev.azure.com/dnceng-public/public/_build/results?buildId=506757&view=ms.vss-test-web.build-test-results-tab&runId=11822648&resultId=111206&paneView=debug) because of moving around the enum ordering.

Thank you for pointing that out. Those asserts I added -- e.g. `assert(!insOptsScalable(opt))` -- are naive to the fact that enum values for SVE and non-SVE `insOpts` now overlap, so it's possible for a valid `insOpts` to be confused with an invalid one, thus triggering the assert. I've removed those asserts, though I'm not sure what other state we can assert instead to ensure `emitDispSveArrangement` is only being used for SVE instructions; the default case for the switch statement will catch these erroneous uses except for when a non-SVE `insOpt` value overlaps with an SVE one, in which case we'll print the wrong suffix (which might be enough of an indicator in asmdiffs that something is wrong?).

kunalspathak commented:
I must say that although the `insOpts` are non-overlapping between SVE and non-SVE, I am little uncomfortable on resetting the enum values because of such bugs or mistakenly passing it around unintentionally resulting in wrong meaning. Just a thought and not for this PR, but I am wondering if we should re-evaluate if we need so many different entries. They are only used while deciding the instruction format in `emitIns_R*` methods (in addition to the asserts) and wondering if we should have a separate version of `emitIns_R*` specific to SVE which propagates the information of `SIMD_SCALAR` vs. `SIMD_VECTOR`, etc. differently to decide the format to pick?

```c++
        case INS_OPTS_SCALABLE_B:
        case INS_OPTS_SCALABLE_WIDE_B:
        case INS_OPTS_SCALABLE_B_WITH_SIMD_SCALAR:
        case INS_OPTS_SCALABLE_B_WITH_SIMD_VECTOR:
        case INS_OPTS_SCALABLE_B_WITH_SCALAR:
        case INS_OPTS_SCALABLE_B_WITH_PREDICATE_MERGE:
            return EA_1BYTE;
```

Of course, the alternative to increase the size of `_idInsOpt` from 6 to 7 might be expensive as that will further reduce our ability for representing small consts in smallest possible `instrDesc`. Can you check the TP cost of doing that in a separate PR?

amanasifkhalid commented:
> wondering if we should have a separate version of emitIns_R* specific to SVE which propagates the information of SIMD_SCALAR vs. SIMD_VECTOR, etc. differently to decide the format to pick?

I like this idea of separating SVE-specific logic out into separate functions. All the new SVE-specific `insOpt` values are a bit confusing, and I imagine we'll end up adding more if we continue our current strategy, so it would be nice to clean these up. Also while not as important, having SVE-specific paths might be slightly faster in terms of TP (smaller switch statements, etc).

> Of course, the alternative to increase the size of _idInsOpt from 6 to 7 might be expensive as that will further reduce our ability for representing small consts in smallest possible instrDesc. Can you check the TP cost of doing that in a separate PR?

Sure thing. I initially tried this strategy, though I ran into some weird build issues from the static `instrDesc` size/offset checks...

a74nh commented:
> Just a thought and not for this PR, but I am wondering if we should re-evaluate if we need so many different entries. They are only used while deciding the instruction format in `emitIns_R*` methods (in addition to the asserts)

This is the crucial part - in some cases, once the group is decided, this information in the `insOpt` is no longer needed. In other cases, the `insOpt` information is still required later (eg - most of the neon instructions).

I'll write a quick PR that introduces a `insGroupOpts` which is an additional optional argument to `emitIns_R_R_R_R()`. This should remove most of the new `insOpt` entries, and means nothing needs to increase in size.

a74nh commented:
> I'll write a quick PR that introduces a `insGroupOpts` which is an additional optional argument to `emitIns_R_R_R_R()`. This should remove most of the new `insOpt` entries, and means nothing needs to increase in size.

See https://github.com/dotnet/runtime/pull/96692

kunalspathak commented:
> See https://github.com/dotnet/runtime/pull/96692

I will wait for this to merge before reviewing this PR.

a74nh commented:
>See https://github.com/dotnet/runtime/pull/96692

This is now merged.

All your `InsOpt` entries should be movable into `InsScalableOpts`. Note, this new enum is not stored in `instr`. So, you can't assert it in the SanityCheck function. It's only used to determine which encoding group to use in `EmitIns_R_R_R_()`

amanasifkhalid committed:
Merge insScalableOpts in

amanasifkhalid commented:
@a74nh thank you for cleaning things up! I've merged your changes in, and updated [this gist](https://gist.github.com/amanasifkhalid/3c9516ea14c1c8e991a3af246fd9d6d6) with the JitDisasm and cstool outputs. They look the same, except for the cstool issue with printing predicate register numbers as hex.

cc @kunalspathak

amanasifkhalid committed:
Fix typo

kunalspathak commented on src/coreclr/jit/instr.h:
```diff
@@ -297,22 +291,33 @@ enum insOpts : unsigned
+
+    // There should be no overlap between non-SVE and SVE values,
+    // so reset value to 1 here
+    INS_OPTS_SCALABLE_B = 1,
```
we don't need resetting anymore. also, you can pretty much revert changes done in `enum insOpts`

kunalspathak commented on src/coreclr/jit/emitarm64.h:
```diff
@@ -1012,6 +1022,17 @@ inline static bool insScalableOptsNone(insScalableOpts sopt)
     return sopt == INS_SCALABLE_OPTS_NONE;
 }
 
+inline static bool insScalableOptsWithPredicatePair(insScalableOpts sopt)
```
looks like we missed the summary header for newer SVE methods we added recently. do you mind adding those?

kunalspathak commented on src/coreclr/jit/emitarm64.cpp:
```diff
@@ -16362,7 +16499,6 @@ size_t emitter::emitOutputInstr(insGroup* ig, instrDesc* id, BYTE** dp)
             {
                 code = insEncodeSveElemsize_dtype(ins, optGetSveElemsize(id->idInsOpt()), code);
             }
-
```
extra line deletion

a74nh reviewed (approved):
LGTM (once other review comments are fixed up)

amanasifkhalid committed:
Feedback

amanasifkhalid commented:
Thank you for the reviews! @kunalspathak I've addressed your feedback.

kunalspathak reviewed (approved):
LGTM
